nfs/target.go

```diff
@@ -202,67 +200,74 @@ func (v *Target) readDirPlus(fh []byte) ([]*EntryPlus, error) {
 		MaxCount   uint32
 	}
 
+	type DirListPlus3 struct {
+		IsSet bool      `xdr:"union"`
+		Entry EntryPlus `xdr:"unioncase=1"`
+	}
+
 	type DirListOK struct {
-		DirAttrs struct {
-			Follows  uint32
-			DirAttrs Fattr
-		}
+		DirAttrs   PostOpAttr
 		CookieVerf uint64
-		Follows    uint32
 	}
 
-	res, err := v.call(&ReadDirPlus3Args{
-		Header: rpc.Header{
-			Rpcvers: 2,
-			Prog:    Nfs3Prog,
-			Vers:    Nfs3Vers,
-			Proc:    NFSProc3ReadDirPlus,
-			Cred:    v.auth,
-			Verf:    rpc.AuthNull,
-		},
-		FH:       fh,
-		DirCount: 512,
-		MaxCount: 4096,
-	})
+	var entries []*EntryPlus
+	for !eof {
+		res, err := v.call(&ReadDirPlus3Args{
+			Header: rpc.Header{
+				Rpcvers: 2,
+				Prog:    Nfs3Prog,
+				Vers:    Nfs3Vers,
+				Proc:    NFSProc3ReadDirPlus,
+				Cred:    v.auth,
+				Verf:    rpc.AuthNull,
+			},
+			FH:         fh,
+			Cookie:     cookie,
+			CookieVerf: cookieVerf,
```

Comment:
So was Cookie and CookieVerf the issue with the NTAP box?  I don't see how the union handling change wouldve fixed the issue.

Reply:
Cookie/CookieVerf were preventing us from getting more than 1 directory entry back. I'm not sure why but the first return always seems to be for `.` and then you get the rest. Could be related to the `FSInfo.dtpref` value, but I couldn't find any expansion on what the preferred multipliers were intended for so left it as is.

The union values were causing most of the problems - for example FSInfo was getting populated with incorrect data because we didn't have the optional FAttr structure - that led to the `wtpref` size being 0 so we wrote 0 bytes at a time, got a short write and then died.

Reply:
Gah. Good point - I'll need to add handling to Write to deal with a wtmax being lower than the data to write.